mipengine/controller/algorithm_specifications.py

```diff
@@ -0,0 +1,165 @@
+    ParameterSpecificationDTO,
+)
+
+
+class InputDataSpecification(BaseModel):
+    label: str
+    desc: str
+    types: List[InputDataType]
+    stattypes: List[InputDataStatType]
+    notblank: bool
+    multiple: bool
+    enumslen: Optional[int] = None
+
+    def convert_to_inputdata_specification_dto(self):
+        # The only difference of the DTO is that it's stattypes is Optional,
+        # due to the fact that the datasets/data_model variables are added.
+        return InputDataSpecificationDTO(
+            label=self.label,
+            desc=self.desc,
+            types=self.types,
+            stattypes=self.stattypes,
+            notblank=self.notblank,
+            multiple=self.multiple,
+            enumslen=self.enumslen,
+        )
+
+
+def get_data_model_input_data_specification_DTO():
+    return InputDataSpecificationDTO(
+        label="Data model of the data.",
+        desc="The data model that the algorithm will run on.",
+        types=[InputDataType.TEXT],
+        notblank=True,
+        multiple=False,
+        stattypes=None,
+        enumslen=None,
+    )
+
+
+def get_datasets_input_data_specification_DTO():
+    return InputDataSpecificationDTO(
+        label="Set of data to use.",
+        desc="The set of data to run the algorithm on.",
+        types=[InputDataType.TEXT],
+        notblank=True,
+        multiple=True,
+        stattypes=None,
+        enumslen=None,
+    )
+
+
+def get_filters_input_data_specification_DTO():
+    return InputDataSpecificationDTO(
+        label="filter on the data.",
+        desc="Features used in my algorithm.",
+        types=[InputDataType.JSONOBJECT],
+        notblank=False,
+        multiple=False,
+        stattypes=None,
+        enumslen=None,
+    )
+
+
+class InputDataSpecifications(BaseModel):
+    x: Optional[InputDataSpecification] = None
+    y: Optional[InputDataSpecification] = None
+
+    def convert_to_inputdata_specifications_dto(self):
+        # In the DTO the datasets, data_model and filter parameters are added from the engine.
+        # These parameters are not added by the algorithm developer.
+        x = self.x.convert_to_inputdata_specification_dto() if self.x else None
+        y = self.y.convert_to_inputdata_specification_dto() if self.y else None
+        return InputDataSpecificationsDTO(
+            x=x,
+            y=y,
+            data_model=get_data_model_input_data_specification_DTO(),
+            datasets=get_datasets_input_data_specification_DTO(),
+            filter=get_filters_input_data_specification_DTO(),
+        )
+
+
+class ParameterSpecification(ParameterSpecificationDTO):
+    # The parameter specification is identical to the DTO
+    pass
+
+
+class AlgorithmSpecification(BaseModel):
+    name: str
+    desc: str
+    label: str
+    enabled: bool
+    inputdata: InputDataSpecifications
+    parameters: Optional[Dict[str, ParameterSpecification]] = None
```

Comment:
When something is `Optional` it automatically gets a `None` default parameter. The `= None` is not needed.

Reply:
Are you sure?

https://pydantic-docs.helpmanual.io/usage/types/#typing-iterables

Search for 
```
    none_or_str: Optional[str] = None
```